apps/src/studio/collisionMaskWalls.js

```diff
@@ -0,0 +1,91 @@
+import Walls from './walls';
+import {imageDataFromURI} from '../imageUtils';
+
+const BYTES_PER_PIXEL = 4;
+const BITS_PER_BYTE = 8;
+
+export default class CollisionMaskWalls extends Walls {
+  constructor(level, skin, drawDebugRect, drawDebugOverlay, width, height, onload) {
+    super(level, skin, drawDebugRect);
+
+    this.width = width;
+    this.height = height;
+    this.drawDebugOverlay = drawDebugOverlay;
+    this.bytesPerRow = Math.ceil(width / BITS_PER_BYTE);
+    this.wallMaps = {};
+    Promise.all(Object.keys(skin.wallMaps).map(mapName => {
+      return imageDataFromURI(skin.wallMaps[mapName].srcUrl).then(imageData => {
+        this.wallMaps[mapName] = {
+          wallMap: this.wallMapFromImageData(imageData),
+          srcUrl: skin.wallMaps[mapName].srcUrl
+        };
+      });
+    })).then(() => {
+      if (onload) {
+        onload();
+      }
+    }).catch(err => {
+      console.error(err);
+    });
+  }
+
+  /**
+   * @override
+   */
+  willRectTouchWall(xCenter, yCenter, collidableWidth, collidableHeight) {
+    if (this.wallMaps[this.wallMapRequested]) {
+      var wallMap = this.wallMaps[this.wallMapRequested].wallMap;
+      this.drawDebugOverlay(this.wallMaps[this.wallMapRequested].srcUrl);
+
+      const yTop = yCenter - collidableHeight / 2;
+      const yBottom = yTop + collidableHeight;
+      const xLeft = xCenter - collidableWidth / 2;
+      const xRight = xLeft + collidableWidth;
+      const firstByteToCheck = Math.floor(xLeft / BITS_PER_BYTE);
+      const lastByteToCheck = Math.floor((xRight - 1) / BITS_PER_BYTE);
+
+      for (let y = yTop; y < yBottom; y++) {
+        const firstByteInRow = this.bytesPerRow * y;
+        for (let x = firstByteToCheck; x <= lastByteToCheck; x++) {
+          if (wallMap[firstByteInRow + x]) {
+            // Check individual bits, starting from either the first bit in the
+            // byte, or the far left of the sprite's collision rect. End with
+            // the last bit or the far right of the collision rect.
+            const start = Math.max(xLeft, x * BITS_PER_BYTE);
+            const end = Math.min(xRight, (x + 1) * BITS_PER_BYTE);
+            const mask = ((1 << (end - start)) - 1) << (start % BITS_PER_BYTE);
+            if (wallMap[firstByteInRow + x] & mask) {
```

Comment:
damn, this is cool. Pretty hard to grok, though; wouldn't hurt to throw a comment in breaking down a bit what's going on, particularly for those who don't know how the data in wallMap is structured.